src/Marten/Events/AsyncProjectionTestingExtensions.cs

```diff
@@ -13,62 +16,76 @@ public static class TestingExtensions
-    /// <param name="timeout"></param>
     public static async Task WaitForNonStaleProjectionDataAsync(this IDocumentStore store, string tenantIdOrDatabaseName, TimeSpan timeout)
     {
         // Assuming there's only one database in this usage
         var database = await store.Storage.FindOrCreateDatabase(tenantIdOrDatabaseName).ConfigureAwait(false);
 
         if (store.Storage is DefaultTenancy)
-
-        await WaitForNonStaleProjectionDataAsync(database, timeout).ConfigureAwait(false);
+            await WaitForNonStaleProjectionDataAsync(database, store.AsyncProjectionsCount(), timeout).ConfigureAwait(false);
     }
 
-    public static async Task WaitForNonStaleProjectionDataAsync(this IMartenDatabase database, TimeSpan timeout)
+    /// <summary>
+    /// Wait for any running async daemon to catch up to the latest event sequence at the time
+    /// </summary>
+    /// <param name="projectionsCount">Will be awaited till all shards have been started before checking if they've caught up with the sequence number</param>
+    /// <exception cref="TimeoutException"></exception>
+    public static async Task WaitForNonStaleProjectionDataAsync(this IMartenDatabase database, int projectionsCount, TimeSpan timeout)
     {
         using var cancellationSource = new CancellationTokenSource();
         cancellationSource.CancelAfter(timeout);
 
-        var initial = await database.FetchEventStoreStatistics(cancellationSource.Token).ConfigureAwait(false);
-        while (initial.EventSequenceNumber == 0 && !cancellationSource.IsCancellationRequested)
+        EventStoreStatistics initial;
+        do
         {
+            initial = await database.FetchEventStoreStatistics(cancellationSource.Token).ConfigureAwait(false);
+            if (initial.EventSequenceNumber > 0 || cancellationSource.IsCancellationRequested)
+                break;
+
             await Task.Delay(100.Milliseconds(), cancellationSource.Token).ConfigureAwait(false);
-        }
+        } while (true);
 
         if (initial.EventSequenceNumber == 0)
         {
-            throw new TimeoutException("No projection or event activity was detected within the timeout span");
+            throw new TimeoutException("No event activity was detected within the timeout span");
         }
 
-        var projections = await database.AllProjectionProgress(cancellationSource.Token).ConfigureAwait(false);
-        while (!projections.All(x => x.Sequence >= initial.EventSequenceNumber) &&
-               !cancellationSource.IsCancellationRequested)
+        IReadOnlyList<ShardState> projections;
+        do
         {
+            projections = await database.AllProjectionProgress(cancellationSource.Token).ConfigureAwait(false);
+            if ((projections.Count >= projectionsCount && projections.All(x => x.Sequence >= initial.EventSequenceNumber))
+                || cancellationSource.IsCancellationRequested)
+                break;
```

Comment:
here are the main problems that i had
1. `projections` were evaluated only once and never updated
2. on _cold_ start `AllProjectionProgress()` returns an empty collection, so `projections.All(...)` evaluation returns false positive result.. that is why i had to explicitly check if all configured async projection shards are running
@baltie this is exactly what you're experiencing